PluginBuilder/Controllers/ApiController.cs

```diff
@@ -106,24 +108,73 @@ public async Task<IActionResult> Download(
         [ModelBinder(typeof(PluginVersionModelBinder))] PluginVersion version)
     {
         await using var conn = await ConnectionFactory.Open();
-        var url = await conn.ExecuteScalarAsync<string?>(
-            "SELECT b.build_info->>'url' FROM versions v " +
+
+        var result = await conn.QueryFirstOrDefaultAsync<(string? Url, long BuildId)>(
+            "SELECT b.build_info->>'url' AS Url, v.build_id AS BuildId " +
+            "FROM versions v " +
             "JOIN builds b ON b.plugin_slug = v.plugin_slug AND b.id = v.build_id " +
             "WHERE v.plugin_slug=@plugin_slug AND v.ver=@version",
             new
             {
                 plugin_slug = pluginSlug.ToString(),
                 version = version.VersionParts
             });
-        if (url is null)
+
+        if (result == default || result.Url is null)
             return NotFound();
 
-        await conn.InsertEvent("Download", new JObject
+        var clientIp = HttpContext.Connection.RemoteIpAddress?.ToString();
+
+        var lastEventType = await conn.GetLastEventTypeForIpAsync(clientIp, pluginSlug);
+        var eventInsertionResponse = await conn.InsertEvent("Download", new JObject
         {
             ["pluginSlug"] = pluginSlug.ToString(),
             ["version"] = version.ToString()
-        });
-        return Redirect(url);
+        }, clientIp, pluginSlug, result.BuildId);
+        if (eventInsertionResponse)
+        {
+            await conn.RecordPluginDownloadStatistics(pluginSlug, "install", version.VersionParts);
+        }
+        else if (lastEventType?.ToLower() == "uninstall")
+        {
+            await conn.RecordPluginDownloadStatistics(pluginSlug, "install", version.VersionParts);
```

Comment:
it is no different from the other branch of the if.

Reply:
Based on Rockstar proposed changes:

Only increment download_stat in versions.download_stat if it’s a fresh insert.

A case scenario that played in my head was that:

"If a user install a plugin, the download_stat would increase, and if the user uninstalls the same plugin the stat would reduce.

If the user reinstalls the same plugin, the download stats should pop back up.. "

So what I am trying to achieve on this endpoint is to check if the last event the user (ip) did on that plugin was an uninstall that would mean the user is trying to re-install the plugin. I am  then increasing the download_stat by one